numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -1127,6 +1127,88 @@ static void
 
 /**end repeat**/
 
+/**begin repeat
+ *
+ * #TOTYPE = DATETIME#
+ * #totype = npy_datetime#
+ */
+
+/**begin repeat1
+ *
+ * #FROMTYPE = BYTE, UBYTE, SHORT, USHORT, INT, UINT, LONG, ULONG,
+ *             LONGLONG, ULONGLONG, DATETIME, TIMEDELTA#
+ * #fromtype = npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+ *             npy_long, npy_ulong, npy_longlong, npy_ulonglong,
+ *             npy_datetime, npy_timedelta#
+ */
+static void
+@FROMTYPE@_to_@TOTYPE@(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const @fromtype@ *ip = input;
+    @totype@ *op = output;
+
+    while (n--) {
+        *op++ = (@totype@)*ip++;
+    }
+}
+/**end repeat1**/
+
+/**begin repeat1
+ *
+ * #FROMTYPE = FLOAT, DOUBLE, LONGDOUBLE#
+ * #fromtype = npy_float, npy_double, npy_longdouble#
+ */
+static void
+@FROMTYPE@_to_@TOTYPE@(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const @fromtype@ *ip = input;
+    @totype@ *op = output;
+
+    while (n--) {
+        @fromtype@ f = *ip++;
+
+        /* Avoid undefined behaviour for NaN -> NaT */
+        if (npy_isnan(f)) {
```

Comment:
An alternative trick you could use here is add a `supports_nat` to the loop template, and then use `#if @supports_nat@` here, which would make the patch shorter.

Reply:
Thanks for the tip, I've used this.